src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -305,6 +306,15 @@ public Void invoke(File f, VirtualChannel channel) throws IOException, Interrupt
         }
     }
 
+    private final Map<String, Object> nodeLocks = new ConcurrentHashMap<>();
+    private static Object lockFor(Node node) {
+        // Avoiding WeakHashMap<Node, T> since Slave overrides hashCode/equals
+        // Avoiding new Object() to prepare for http://cr.openjdk.java.net/~briangoetz/valhalla/sov/02-object-model.html
+        // Avoiding new String(…) because static analyzers complain
+        // Could use anything but hoping that a future JVM enhanced thread dumps to display monitors of type String
+        return ExtensionList.lookupSingleton(WorkspaceLocatorImpl.class).nodeLocks.computeIfAbsent(node.getNodeName(), name -> new StringBuilder("WorkspaceLocatorImpl lock for ").append(name).toString());
```

Comment:
this is safe but will leak for cloud agents that use unique node names.

Reply:
True, but only leaks a shortish `String`.

We could also use your original suggestion of `node.getNodeName().intern()`…which would also introduce a similar small leak. No strong opinion.

Reply:
I'm trying to recall where we solved this last (without guava of course!)

Reply:
can't find it.

https://github.com/apache/avro/blob/d5d1cd70e920f45468bf696d3668b496a9adbab3/lang/java/avro/src/main/java/org/apache/avro/util/WeakIdentityHashMap.java is should be license compatible which we may be able to could copy and use here (and then pull into Jenkins core in the future).  WDYT?

String interning was moved from permgen in 1.7 so should be better, but there is the question of thrashing the pool.

1) use guava from core 
2) introduce a small leak
3) use String interning
4) use third party code (either an extra dep or using some appropriately licensed code that is copied into the codebase)
5) write our own WeakIdentityHashMap

I think arguments could be made either way, I guess we just need a consensus on this.

I generally don't like introducing leaks, and am against introducing guava.
Making heavy use of the intern pool would have some unknowns (not sure how much else makes use of the pool to know what implications that would have), so is a bit of an unknown.
So 4 or 5 seem slightly better, and not sure we should write our own if there are appropriatley licenced ones we could use.

WDYT?

Reply:
> use guava from core

`CacheBuilder` with `weakKeys`? We already use this in core and Pipeline (https://github.com/jenkinsci/jenkins/blob/3f15ad20985066661158ee1bb0b3255e000b3081/core/src/main/java/hudson/model/Queue.java#L221 + https://github.com/jenkinsci/workflow-support-plugin/blob/173378ee6f1178b73f4ad1c18d0250401455db6a/src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java#L74) so I do not see why not. This would be my preference.

Reply:
> `CacheBuilder` with `weakKeys`

yes.

if you are happy with that then lets go with it.  not part of any serialised form so it at least safe from that PoV.